crates/llm-local/src/llama.rs

```diff
@@ -0,0 +1,209 @@
+        let eos_token_id = config.clone().eos_token_id.or_else(|| {
+            tokenizer
+                .token_to_id(EOS_TOKEN)
+                .map(llama::LlamaEosToks::Single)
+        });
+
+        let mut tokens = tokenizer
+            .encode(prompt, true)
+            .map_err(|e| anyhow!(e.to_string()))?
+            .get_ids()
+            .to_vec();
+        let mut tokenizer = token_output_stream::TokenOutputStream::new(tokenizer);
+        let mut rng = rand::rngs::StdRng::from_entropy();
+
+        let mut logits_processor = {
+            let temperature = params.temperature;
+            let sampling = if temperature <= 0. {
+                Sampling::ArgMax
+            } else {
+                Sampling::TopKThenTopP {
+                    k: params.top_k as usize,
+                    p: params.top_p as f64,
+                    temperature: params.temperature as f64,
+                }
+            };
+            LogitsProcessor::from_sampling(rng.next_u64(), sampling)
+        };
+
+        let mut index_pos = 0;
+        let mut tokens_generated = 0;
+        let mut output_text: String = String::default();
+
+        for index in 0..params.max_tokens {
+            let (context_size, context_index) = if self.cache.use_kv_cache && index > 0 {
+                (1, index_pos)
+            } else {
+                (tokens.len(), 0)
+            };
+            let ctxt = &tokens[tokens.len().saturating_sub(context_size)..];
+            let input = Tensor::new(ctxt, &self.device)?.unsqueeze(0)?;
+            let logits = model.forward(&input, context_index, &mut cache)?;
+            let logits = logits.squeeze(0)?;
+            let logits = if params.repeat_penalty == 1. {
+                logits
+            } else {
+                let start_at = tokens
+                    .len()
+                    .saturating_sub(params.repeat_penalty_last_n_token_count as usize);
+                candle_transformers::utils::apply_repeat_penalty(
+                    &logits,
+                    params.repeat_penalty,
+                    &tokens[start_at..],
+                )?
+            };
+            index_pos += ctxt.len();
+
+            let next_token = logits_processor.sample(&logits)?;
+            tokens_generated += 1;
+            tokens.push(next_token);
+
+            // Validate if we have reached the end of the token(s)
+            match eos_token_id {
+                Some(llama::LlamaEosToks::Single(eos_tok_id)) if next_token == eos_tok_id => {
+                    break;
+                }
+                Some(llama::LlamaEosToks::Multiple(ref eos_ids))
+                    if eos_ids.contains(&next_token) =>
+                {
+                    break;
+                }
+                _ => (),
+            }
+            // Decode the token and add it to the output.
+            if let Some(t) = tokenizer.next_token(next_token)? {
+                print!("{}", t);
```

Comment:
Do we want to print this?

Reply:
That was used for some debugging I will remove it.